kaitaistruct.py

```diff
@@ -15,51 +20,171 @@
-    def __init__(self, stream):
-        self._io = stream
+class _KaitaiStruct_:
+    __slots__ = ("_io", "_parent", "_root")
+    def __init__(self, _io:"KaitaiStream", _parent:typing.Optiohal["_KaitaiStruct"]=None, _root:typing.Optiohal["_KaitaiStruct"]=None):
+        self._io = _io
+        self._parent = _parent
+        self._root = _root if _root else self
+
+
+class KaitaiParser(ABC):
+    __slots__ = () # in fact this class not meant to be instantiated
+
+    @abstractmethod
+    @classmethod
+    def parse(cls, struct:_KaitaiStruct_, io:"KaitaiStream"):
+        raise NotlmplementedError()
+
+
+class _KaitaiStruct(_KaitaiStruct_):
+    __slots__ = ()
+    _parser = None  # :KaitaiParser
+
+    def _read(self):
+        self.__class__._parser.parse(self, self._io)
+
+
+class KaitaiStruct(_KaitaiStruct):
+    __slots__ = ("_shouldExit",)
+    def __init__(self, io: "KaitaiStream"):
+        super.__init__(io)
+        self._shouldExit = False
 
     def __enter__(self):
+        self._shouldExit = not stream.is_entered
+        if self._shouldExit:
+            self._io.__enter__()
         return self
 
     def __exit__(self, *args, **kwargs):
-        self.close()
-
-    def close(self):
-        self._io.close()
+        if self.shouldExit:
+            self._io.__exit__(*args, **kwargs)
 
     @classmethod
-    def from_file(cls, filename):
-        f = open(filename, 'rb')
-        try:
-            return cls(KaitaiStream(f))
-        except Exception:
-            # close file descriptor, then reraise the exception
-            f.close()
-            raise
+    def from_any(cls, o: typing.Union[Path, str]):
+        with KaitaiStream(o) as io:
+            s = cls(io)
+            s._read()
+            return s
 
     @classmethod
-    def from_bytes(cls, buf):
-        return cls(KaitaiStream(BytesIO(buf)))
+    def from_file(cls, file: typing.Union[Path, str], use_mmap:bool=True):
+        return cls.from_any(file, use_mmap=use_mmap)
 
-    @classmethod
-    def from_io(cls, io):
-        return cls(KaitaiStream(io))
 
+class IKaitaiDownStream(ABC):
+    __slots__ =("_io",)
 
-class KaitaiStream(object):
-    def __init__(self, io):
-        self._io = io
-        self.align_to_byte()
+    def __init__(self, _io: typing.Any):
+        self._io = _io
+
+    @abstractmethod
+    @property
+    def is_entered(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def __enter__(self):
+        raise NotImplementedError()
+
+    def __exit__(self, *args, **kwargs):
+        if self.is_entered():
+            self._io.__exit__(*args, **kwargs)
+            self._io = None
+
+    def __del__(self):
+        self.__exit__(None, None)
+
+
+class KaitaiFileSyscallDownStream(IKaitaiDownStream):
+    __slots__ = ()
+    def __init__(self, path: typing.Union[Path, str]):
+        super().__init__(Path(path))
+
+    @property
+    def is_entered(self):
+        return isinstance(self._io, _IOBase)
```

Comment:
This is almost guaranteed to not work, because most built-in IO streams don't really extend `_io._IOBase`. You need to use `io.IOBase` instead - that will work, because it is an ABC and has all built-in IO stream classes registered as virtual subclasses.

Reply:
>don't really extend _io._IOBase

At the time I have first written the code (python 3.4, the latest version available on Win XP) for this PR `mro()` output had `_io._IOBase`.

Reply:
Ah ok, I remembered that incorrectly - apparently the streams from `_io` do actually extend `_io._IOBase`.

In any case, `_io` is not a documented module, so there's no guarantee that `_io._IOBase` will be used that way in the future. `io.IOBase` is documented and does exactly the same thing here, so you should really use that instead.